File: Android/NativeKeyboard/NativeKeyboard/src/main/java/com/jeroenvanpienbroek/nativekeyboard/ThreadsafeQueue.java
package com.jeroenvanpienbroek.nativekeyboard;

import java.util.Queue;
import java.util.concurrent.LinkedBlockingQueue;

/** Class that wraps Queue in a threadsafe way */
public class ThreadsafeQueue<T>
{
    private Queue<T> queue;
    private Object queueLock = new Object();

    public ThreadsafeQueue()
    {
        this.queue = new LinkedBlockingQueue<T>();
    }

    public void enqueue(T item)
    {
        synchronized (queueLock)
        {
            this.queue.add(item);
        }
    }

    public T dequeue()
    {
        synchronized (queueLock)
        {
            return this.queue.poll();
        }
    }

    public boolean contains(T item)
    {
        synchronized (queueLock)
        {
            return this.queue.contains(item);
        }
    }

    public int getCount()
    {
        synchronized (queueLock)
        {
            return this.queue.size();
        }
    }

    public void clear()
    {
        synchronized (queueLock)
        {
            this.queue.clear();
        }
    }
}

File: UWP/NativeKeyboard/NativeKeyboard/ThreadsafeQueue.cs
﻿using System.Collections.Generic;

namespace NativeKeyboardUWP
{
	/// <summary>Class that wraps Queue in a threadsafe way</summary>
	public class ThreadsafeQueue<T>
	{
		private Queue<T> queue;
		private object queueLock = new object();

		public ThreadsafeQueue()
		{
			this.queue = new Queue<T>();
		}

		public void Enqueue(T item)
		{
			lock(queueLock)
			{
				this.queue.Enqueue(item);
			}
		}

		public T Dequeue()
		{
			lock(queueLock)
			{
				return this.queue.Dequeue();
			}
		}

		public bool Contains(T item)
		{
			lock(queueLock)
			{
				return this.queue.Contains(item);
			}
		}

		public int Count
		{
			get
			{
				lock(queueLock)
				{
					return this.queue.Count;
				}
			}
		}

		public void Clear()
		{
			lock(queueLock)
			{
				this.queue.Clear();
			}
		}

		public IEnumerator<T> GetEnumerator()
		{
			lock(queueLock)
			{
				return this.queue.GetEnumerator();
			}
		}
	}
}
